### PR TITLE
Add dedup processor

### PR DESCRIPTION
**Pull request by igomura**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

**Comment by igomura** [comment]
Hi, my I have review for my PR? Is that something you would consider for 1.13.1?

**Comment by danielnelson** [comment]
> Is that something you would consider for 1.13.1?

We are only doing bug fixes for 1.13.x releases, but we will try for 1.14.

File: plugins/processors/dedup/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+### Configuration
+
+```toml
+[[processors.dedup]]
+  ## Maximum time to suppress output
+  dedup_interval = "600s"
+
+  ## Maximum time to keep cached metric without update
+  evict_interval = "1h"
```
Comment (danielnelson) [inline plugins/processors/dedup/README.md]: I don't understand why we need two durations, could these be combined?

File: plugins/processors/dedup/dedup.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	d.Cache = keep
+}
+
+func (d *Dedup) Apply(metrics ...telegraf.Metric) []telegraf.Metric {
+	for idx, metric := range metrics {
+		id := metric.HashID()
+		if m, ok := d.Cache[id]; ok {
+			// compare all fields values
+			sameValue := true
+			for k, v := range metric.Fields() {
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: Use `metric.FieldList()` here since it is much faster.

File: plugins/processors/dedup/dedup.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+				if m.Fields()[k] != v {
+					log.Printf("D! [processors.dedup]: metric value has changed: %s", metric.Name())
+					sameValue = false
+				}
+			}
+			// value has not changed and cache is not expired
+			if sameValue && time.Since(m.Time()) < d.DedupInterval.Duration {
+				// Deduplicate this metric
+				log.Printf("D! [processors.dedup]: suppress metric: %s", metric.Name())
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: Can you remove all the logging messages, they are too detailed for our current logging system.

File: plugins/processors/dedup/dedup.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+				metrics = remove(metrics, idx)
+				continue
+			}
+		}
+		log.Printf("D! [processors.dedup]: update cached metric: %s", metric.Name())
+		d.Cache[id] = metric.Copy()
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: Call metric.Accept() on the copy, this is important in case the cached metric is a "tracking metrics"

File: plugins/processors/dedup/dedup.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+// Remove single item from slice
+func remove(slice []telegraf.Metric, i int) []telegraf.Metric {
+	slice[len(slice)-1], slice[i] = slice[i], slice[len(slice)-1]
+	return slice[:len(slice)-1]
+}
+
+// Remove expired items from cache
+func (d *Dedup) cleanup() {
+	keep := make(map[uint64]telegraf.Metric, 0)
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: Since the processors get called for every metric, I think we should limit how often this is called and reuse the cache map.

File: scripts/stretch.docker [hunk 6]
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.11.0 as builder
+FROM golang:1.12.0 as builder
```
Comment (danielnelson) [inline scripts/stretch.docker]: Revert unrelated change

File: .gitignore [hunk 7]
```diff
@@ -3,3 +3,4 @@
 /telegraf.exe
 /telegraf.gz
 /vendor
+.idea
```
Comment (danielnelson) [inline .gitignore]: Please revert this change, you can add it to your global Git `core.excludesfile`

File: plugins/processors/dedup/README.md (hunk 1 above)
Comment (igomura) [reply plugins/processors/dedup/README.md]: You are right. I will combine them

**Comment by igomura** [comment]
Thanks for review. This is my "Hello World" in go. I appreciate any suggestion to make this module better.

File: plugins/processors/dedup/dedup.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+func (d *Dedup) Apply(metrics ...telegraf.Metric) []telegraf.Metric {
+	for idx, metric := range metrics {
+		id := metric.HashID()
+		// check if metric is already in cache. Otherwise save it in cache
+		if m, ok := d.Cache[id]; ok {
+			// if cache is not expired then check values. Otherwise save it in cache
+			if time.Since(m.Time()) < d.DedupInterval.Duration {
+				for _, field := range metric.FieldList() {
+					// if same value then drop it. Otherwise save it in cache
+					if m.Fields()[field.Key] == field.Value {
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: Use `m.GetField(field.Key)` here instead of `m.Fields()`.  The Fields() method is very slow, hopefully we can remove it soon to avoid confusion.

File: plugins/processors/dedup/dedup.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+		if time.Since(metric.Time()) < d.DedupInterval.Duration {
+			keep[id] = metric
+		}
+	}
+	d.Cache = keep
+}
+
+// Save item to cache
+func (d *Dedup) save(metric telegraf.Metric) {
+	id := metric.HashID()
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: Let's pass in the `id` we obtained from line 69, to avoid recalculating it.

File: plugins/processors/dedup/dedup.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+			// if cache is not expired then check values. Otherwise save it in cache
+			if time.Since(m.Time()) < d.DedupInterval.Duration {
+				for _, field := range metric.FieldList() {
+					// if same value then drop it. Otherwise save it in cache
+					if m.Fields()[field.Key] == field.Value {
+						metrics = remove(metrics, idx)
+					} else {
+						d.save(metric)
+					}
+					continue
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: Since this is the last line of a for loop `continue` won't have any effect, we should remove it.

File: plugins/processors/dedup/dedup.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+		// check if metric is already in cache. Otherwise save it in cache
+		if m, ok := d.Cache[id]; ok {
+			// if cache is not expired then check values. Otherwise save it in cache
+			if time.Since(m.Time()) < d.DedupInterval.Duration {
+				for _, field := range metric.FieldList() {
+					// if same value then drop it. Otherwise save it in cache
+					if m.Fields()[field.Key] == field.Value {
+						metrics = remove(metrics, idx)
+					} else {
+						d.save(metric)
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: If we are saving the metric in the cache, I think we can skip looking at the other fields.

File: plugins/processors/dedup/dedup.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+					}
+					continue
+				}
+			} else {
+				d.save(metric)
+			}
+		} else {
+			d.save(metric)
+		}
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup.go]: It is a little hard to match these else statements up with the conditions above.  I'd flip the cases around like this:
```
if m, ok := d.Cache[id]; !ok {
    d.save(metric)
    continue
}

// put the "ok" case here
```

File: plugins/processors/dedup/dedup_test.go [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+	assertCacheRefresh(t, &deduplicate, source)
+	assertMetricPassed(t, target, source)
+}
+
+func TestSuppressRepeatedValue(t *testing.T) {
+	deduplicate := createProcessor()
+	source := createMetric("m1", 1)
+	target := deduplicate.Apply(source)
+	// wait less than deduplication interval
+	time.Sleep(1 * time.Second)
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup_test.go]: We shouldn't actually sleep in the tests, since it will result in a race condition and makes the tests slow.  How we usually handle this is to allow the current time to be set in the test.  Here are the basic steps to set this up:
- define a function type for getting the time: `type TimeFunc func() time.Time`
- add a TimeFunc field on the processor struct
- initialize the new field to time.Now
- in the tests, you can swap the function out to change the time.

File: plugins/processors/dedup/dedup_test.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+		FlushTime:     time.Now(),
+		Cache:         make(map[uint64]telegraf.Metric),
+	}
+}
+
+func assertCacheRefresh(t *testing.T, proc *Dedup, item telegraf.Metric) {
+	id := item.HashID()
+	name := item.Name()
+	// cache is not empty
+	assert.NotEqual(t, 0, len(proc.Cache))
```
Comment (danielnelson) [inline plugins/processors/dedup/dedup_test.go]: Use `require` instead of `assert` everywhere you can, the difference is that require will halt the test and with assert it will keep going:
```
require.NotEqual(t, 0, len(proc.Cache))
```

File: plugins/processors/dedup/dedup.go (hunk 10 above)
Comment (igomura) [reply plugins/processors/dedup/dedup.go]: This is attempt to skip looking other fields metric may have. So, it may have affect.

File: plugins/processors/dedup/dedup.go (hunk 11 above)
Comment (igomura) [reply plugins/processors/dedup/dedup.go]: Yep. this is what line 81 for

**Comment by igomura** [comment]
Let me know if you would like me to squash commits at the end
